File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/nurikabe/domain/NurikabeGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.F2;

public class NurikabeGame extends CellsGame<NurikabeGame, NurikabeGameMove, NurikabeGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };
    public static Position offset2[] = {
            new Position(0, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(1, 1),
    };

    public Map<Position, Integer> pos2hint = new HashMap<>();

    public NurikabeGame(List<String> layout, GameInterface<NurikabeGame, NurikabeGameMove, NurikabeGameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        size = new Position(layout.size(), layout.get(0).length());
        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                char ch = str.charAt(c);
                if (ch >= '0' && ch <= '9') {
                    int n = ch - '0';
                    pos2hint.put(p, n);
                }
            }
        }
        NurikabeGameState state = new NurikabeGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(NurikabeGameMove move, F2<NurikabeGameState, NurikabeGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        NurikabeGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(NurikabeGameMove move) {
        return changeObject(move, NurikabeGameState::switchObject);
    }

    public boolean setObject(NurikabeGameMove move) {
        return changeObject(move, NurikabeGameState::setObject);
    }

    public NurikabeObject getObject(Position p) {
        return state().get(p);
    }

    public NurikabeObject getObject(int row, int col) {
        return state().get(row, col);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/nurikabe/NurikabeGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.nurikabe

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class NurikabeGame(layout: List<String>, gi: GameInterface<NurikabeGame, NurikabeGameMove, NurikabeGameState>, gdi: GameDocumentInterface) : CellsGame<NurikabeGame, NurikabeGameMove, NurikabeGameState>(gi, gdi) {
    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1)
        )
        var offset2 = arrayOf(
            Position(0, 0),
            Position(0, 1),
            Position(1, 0),
            Position(1, 1)
        )
    }

    var pos2hint = mutableMapOf<Position, Int>()

    init {
        size = Position(layout.size, layout[0].length)
        for (r in 0 until rows) {
            val str = layout[r]
            for (c in 0 until cols) {
                val p = Position(r, c)
                val ch = str[c]
                if (ch in '0'..'9') {
                    val n = ch - '0'
                    pos2hint[p] = n
                }
            }
        }
        val state = NurikabeGameState(this)
        states.add(state)
        levelInitilized(state)
    }

    private fun changeObject(move: NurikabeGameMove, f: (NurikabeGameState, NurikabeGameMove) -> Boolean): Boolean {
        if (canRedo) {
            states.subList(stateIndex + 1, states.size).clear()
            moves.subList(stateIndex, states.size).clear()
        }
        val state = cloner.deepClone(currentState)
        val changed = f(state, move)
        if (changed) {
            states.add(state)
            stateIndex++
            moves.add(move)
            moveAdded(move)
            levelUpdated(states[stateIndex - 1], state)
        }
        return changed
    }

    fun switchObject(move: NurikabeGameMove) = changeObject(move, NurikabeGameState::switchObject)
    fun setObject(move: NurikabeGameMove) = changeObject(move, NurikabeGameState::setObject)

    fun getObject(p: Position) = currentState[p]
    fun getObject(row: Int, col: Int) = currentState[row, col]
}
